spinalcordtoolbox/csa_pmj.py

```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python
+# -*- coding: utf-8
+# Functions to get distance from PMJ for processing segmentation data
+# Author: Sandrine Bédard
+import logging
+import numpy as np
+
+from spinalcordtoolbox.image import Image
+from spinalcordtoolbox.centerline.core import get_centerline
+from spinalcordtoolbox.resampling import resample_nib
+
+logger = logging.getLogger(__name__)
+
+NEAR_ZERO_THRESHOLD = 1e-6
+
+
+def get_slices_for_pmj_distance(segmentation, pmj, distance, extent, param_centerline=None, verbose=1):
+    """
+    Compute distance from PMJ projection on centerline for all the centerline.
+    Generate mask from segmentation of the slices used to process segmentation data corresponding to a distance from PMJ projection.
+    :param segmentation: input segmentation. Could be either an Image or a file name.
+    :param pmj: label of PMJ.
+    :param distance: float: Distance from PMJ.
+    :param param_centerline: see centerline.core.ParamCenterline()
+    :param verbose:
+    :return mask:
+    :return slices:
+
+    """
+    # TODO: loop through distance (when multiple distances are given)
+    im_seg = Image(segmentation).change_orientation('RPI')
+    im_pmj = Image(pmj).change_orientation('RPI')
+    nx, ny, nz, nt, px, py, pz, pt = im_seg.dim
+    pr = min([px, py])
+    im_segr = resample_nib(im_seg, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    im_pmjr = resample_nib(im_pmj, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    data_pmj = im_pmjr.data
+    # Update dimensions from resampled image
+    nx, ny, nz, nt, px, py, pz, pt = im_segr.dim
+
+    # Extract min and max index in Z direction
+    data_seg = im_segr.data
+    X, Y, Z = (data_seg > NEAR_ZERO_THRESHOLD).nonzero()
+    min_z_index, max_z_index = min(Z), max(Z)
+
+    # Compute the spinal cord centerline based on the spinal cord segmentation
+    _, arr_ctl, arr_ctl_der, fit_results = get_centerline(im_segr, param=param_centerline, verbose=verbose)
+
+    # Get tagent vector at the top of the centerline | Maybe take a fiew slices to get the vector ?
+    iz = max_z_index
+    tangent_vect = np.array([arr_ctl_der[0][iz - min_z_index] * px, arr_ctl_der[1][iz - min_z_index] * py, pz])
+    # Normalize vector by its L2 norm
+    tangent_vect = tangent_vect / np.linalg.norm(tangent_vect)
+    plane_pt = arr_ctl[:, -1] + [0, 0, min_z_index]
+    line_pt = np.argwhere(data_pmj != 0)[0]  # list(im_pmjr.getNonZeroCoordinates()[0])[:-1]
+
+    # Get intersection of plane passing through PMJ and continuation of centerline
+    init_point = get_intersection_plane_line(plane_pt, tangent_vect, line_pt, tangent_vect)
+    # Get the distance from PMJ for all slices of the centerline
+    arr_length = get_distance_from_pmj(arr_ctl, init_point, px, py, pz)
+
+    # Check if distance is out of bound
+    if distance > arr_length[0][0]:
```

Comment:
you might also want to check if `distance < arr_length[0][-1]`

Reply:
done in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3429/commits/6bccae41cf8296490437e0adf5f2e03a7a084cc6